File: src/engine/Component/SoundSource.jl
module SoundSourceModule
    using ..Component.JulGame
    import ..Component
    
    export SoundSource
    struct SoundSource
        channel::Int
        isMusic::Bool
        path::String
        playOnStart::Bool
        volume::Int
    end

    export InternalSoundSource
    mutable struct InternalSoundSource
        path::String
        isMusic::Bool
        channel::Int
        isPlaying::Bool
        playOnStart::Bool
        parent::Any
        sound::Union{Ptr{Nothing}, Ptr{SDL2.LibSDL2._Mix_Music}, Ptr{SDL2.LibSDL2.Mix_Chunk}}
        volume::Int

        # Music
        function InternalSoundSource(parent::Any, path::String, channel::Int = -1, volume::Int = -1, isMusic::Bool = false, playOnStart::Bool = false)
            this = new()

            SDL2.SDL_ClearError()
            fullPath = joinpath(BasePath, "assets", "sounds", path)
            if length(path) < 1
                sound = C_NULL    
            else
                sound = load_sound_sdl(path, isMusic)
            end
            error = unsafe_string(SDL2.SDL_GetError())

            if (sound == C_NULL || !isempty(error)) && length(path) > 0
                println(fullPath)
                error("Error loading file at $path. SDL Error: $(error)")
                SDL2.SDL_ClearError()
            end
            
            # Convert channel and volume to Int32
            isMusic ? SDL2.Mix_VolumeMusic(Math.TypeConversions.safe_int32_convert(clamp(volume, 0, 128))) : SDL2.Mix_Volume(channel, Math.TypeConversions.safe_int32_convert(clamp(volume, 0, 128)))

            this.channel = channel
            this.isMusic = isMusic
            this.parent = parent
            this.path = path
            this.sound = sound
            this.volume = volume
            this.isPlaying = false
            this.playOnStart = playOnStart

            return this
        end
    end

    function Component.toggle_sound(this::InternalSoundSource, loops = 0)
        @debug("Toggling sound from $(this.path), isMusic: $(this.isMusic), loops: $(loops)")
        try
            if this.isMusic
                if SDL2.Mix_PlayingMusic() == 0
                    SDL2.Mix_PlayMusic( this.sound, Math.TypeConversions.safe_int32_convert(-1) )
                    this.isPlaying = true
                else
                    if SDL2.Mix_PausedMusic() == 1 
                        SDL2.Mix_ResumeMusic()
                        this.isPlaying = true
                    else
                        SDL2.Mix_PauseMusic()
                        this.isPlaying = false
                    end
                end
            else
                if SDL2.Mix_PlayChannel(Math.TypeConversions.safe_int32_convert(this.channel), this.sound, Math.TypeConversions.safe_int32_convert(loops)) == -1
                    @error "Error playing channel $(unsafe_string(SDL2.SDL_GetError()))"
                    throw(e)
                end
            end
        catch e
            @error "Error in toggle_sound" exception=(e, catch_backtrace())
        end
    end
    
    function Component.stop_music(this::InternalSoundSource)
        @debug("Stopping music from $(this.path)")
        SDL2.Mix_HaltMusic()
    end

    function Component.load_sound(this::InternalSoundSource, soundPath::String, isMusic::Bool)
        @debug("Loading sound from $(soundPath), isMusic: $(isMusic)")
        this.isMusic = isMusic
        SDL2.SDL_ClearError()
        this.sound = load_sound_sdl(soundPath, isMusic)
        error = unsafe_string(SDL2.SDL_GetError())
        if !isempty(error)
            println(string("Couldn't open sound! SDL Error: ", error))
            SDL2.SDL_ClearError()
            this.sound = C_NULL
            return
        end
        this.path = soundPath
    end

    function load_sound_sdl(soundPath::String, isMusic::Bool)
        @debug("Loading sound from $(soundPath), isMusic: $(isMusic)")
        if haskey(JulGame.AUDIO_CACHE, get_comma_separated_path(soundPath))
            raw_data = JulGame.AUDIO_CACHE[get_comma_separated_path(soundPath)]
            rw = SDL2.SDL_RWFromConstMem(pointer(raw_data), length(raw_data))
            if rw != C_NULL
                @debug("loading sound from cache")
                @debug("comma separated path: ", get_comma_separated_path(soundPath))
                return isMusic ? SDL2.Mix_LoadMUS_RW(rw, 1) : SDL2.Mix_LoadWAV_RW(rw, 1)
            end
        end
        @debug "Loading sound from disk, there are $(length(JulGame.AUDIO_CACHE)) sounds in cache"

        fullPath = joinpath(BasePath, "assets", "sounds", soundPath)
        return isMusic ? SDL2.Mix_LoadMUS(fullPath) : SDL2.Mix_LoadWAV(fullPath)
    end

    function get_comma_separated_path(path::String)
        # Normalize the path to use forward slashes
        normalized_path = replace(path, '\\' => '/')
        
        # Split the path into components
        parts = split(normalized_path, '/')
        
        result = join(parts[1:end], ",")
    
        return result  
    end

    function Component.unload_sound(this::InternalSoundSource)
        @debug("Unloading sound from $(this.path), isMusic: $(this.isMusic)")
        if this.isMusic
            SDL2.Mix_FreeMusic(this.sound)
        else
            SDL2.Mix_FreeChunk(this.sound)
        end
        this.sound = C_NULL
    end

    function Component.set_volume(this::InternalSoundSource, volume::Int = 128, channel::Int = -1)
        @debug("Setting volume for $(this.path), isMusic: $(this.isMusic), volume: $(volume), channel: $(channel)")
        # Convert volume to Int32 for SDL
        this.volume = clamp(volume, 0, 128)
        this.channel = clamp(channel, -1, 128)
        @info "Setting volume for $(this.path), isMusic: $(this.isMusic), volume: $(this.volume), channel: $(this.channel)"
        this.isMusic ? SDL2.Mix_VolumeMusic(Math.TypeConversions.safe_int32_convert(this.volume)) : SDL2.Mix_Volume(this.channel, Math.TypeConversions.safe_int32_convert(this.volume))
    end

    function Component.play(this::InternalSoundSource, loops::Int = 0)
        # Convert loops to Int32
        loops = Math.TypeConversions.safe_int32_convert(loops)
        
        @debug("Playing sound from $(this.path), isMusic: $(this.isMusic), channel: $(this.channel), loops: $(loops)")
        if this.isMusic
            SDL2.Mix_PlayMusic(this.sound, -1)
        else
            SDL2.Mix_PlayChannel(this.channel, this.sound, loops)
        end
    end

    function set_master_volume(volume::Int)
        # Convert volume to Int32 and clamp between 0 and 128
        @info("Setting master volume to $(volume)")
        volume = Math.TypeConversions.safe_int32_convert(clamp(volume, 0, 128))
        SDL2.Mix_MasterVolume(volume)
    end

    function Component.duplicate(this::InternalSoundSource, parent::Any)
        @debug("Duplicating sound from $(this.path), isMusic: $(this.isMusic), channel: $(this.channel), volume: $(this.volume), playOnStart: $(this.playOnStart)")
        newSoundSource = InternalSoundSource(parent, this.path, this.channel, this.volume, this.isMusic, this.playOnStart)
        newSoundSource.isPlaying = this.isPlaying
        return newSoundSource
    end
end

File: ts/_generated/src/engine/Component/SoundSource.ts
export {}
import { clamp, haskey, joinpath, mixVolume, pointer, unsafe_string } from "../../../../src/engine/core/juliaHelpers";


    // using ..Component.JulGame
    // import ..Component
    // include(joinpath(@__DIR__, "SoundSource", "constants.jl"))
    
    
    class SoundSource {
        channel: number
        isMusic: boolean
        path: string
        playOnStart: boolean
        volume: number
    }

    
    class InternalSoundSource {
        path: string
        isMusic: boolean
        channel: number
        isPlaying: boolean
        playOnStart: boolean
        parent: any
        sound: null | any | any
        volume: number

        // Music
        constructor(parent: any, path: string, channel: number = -1, volume: number = -1, isMusic: boolean = false, playOnStart: boolean = false) {
            

            let sound = null

            // Convert channel and volume to Int32
            isMusic ? (globalThis as any).JulGameSdl.glue_Mix_VolumeMusic(mixVolume(volume)) : (globalThis as any).JulGameSdl.glue_Mix_Volume(channel, mixVolume(volume))

            this.channel = channel
            this.isMusic = isMusic
            this.parent = parent
            this.path = path
            this.sound = sound
            this.volume = volume
            this.isPlaying = false
            this.playOnStart = playOnStart

        }
    }

    function Component_toggle_sound(self: InternalSoundSource | null, loops = 0) {
        if (self === null) {
            console.warn("toggle_sound: SoundSource is null")
            return
        }
        console.debug(`toggle_sound: Toggling sound from ${self.path}, isMusic: ${self.isMusic}, loops: ${loops}`)
        try {
            if (self.isMusic) {
                if ((globalThis as any).JulGameSdl.glue_Mix_PlayingMusic() == 0) {
                    (globalThis as any).JulGameSdl.glue_Mix_VolumeMusic(mixVolume(self.volume));
                    (globalThis as any).JulGameSdl.glue_Mix_PlayMusic( self.sound, -1 )
                    self.isPlaying = true
                } else {
                    if ((globalThis as any).JulGameSdl.glue_Mix_PausedMusic() == 1) {
                        (globalThis as any).JulGameSdl.glue_Mix_ResumeMusic()
                        self.isPlaying = true
                    } else {
                        (globalThis as any).JulGameSdl.glue_Mix_PauseMusic()
                        self.isPlaying = false
                    }
                }
            } else {
                (globalThis as any).JulGameSdl.glue_Mix_Volume(self.channel, mixVolume(self.volume));
                if ((globalThis as any).JulGameSdl.glue_Mix_PlayChannel(self.channel, self.sound, loops) == -1) {
                    console.error(`toggle_sound: Error playing channel ${unsafe_string((globalThis as any).JulGameSdl.glue_SDL_GetError())}`)

                }
            }
        } catch (e) {
            console.error("toggle_sound: Error in toggle_sound")
        }
    }
    
    function Component_stop_music(self: InternalSoundSource) {
        console.debug(`stop_music: Stopping music from ${self.path}`);
        (globalThis as any).JulGameSdl.glue_Mix_HaltMusic()
    }

    function Component_load_sound(self: InternalSoundSource, soundPath: string, isMusic: boolean) {
        console.debug(`load_sound: Loading sound from ${soundPath}, isMusic: ${isMusic}`)
        self.isMusic = isMusic;
        (globalThis as any).JulGameSdl.glue_SDL_ClearError()
        self.sound = load_sound_sdl(soundPath, isMusic)
        let error = unsafe_string((globalThis as any).JulGameSdl.glue_SDL_GetError())
        if (error.length > 0) {
            console.log(["Couldn't open sound! SDL Error: ", error].join(""));
            (globalThis as any).JulGameSdl.glue_SDL_ClearError()
            self.sound = null
            return
        }
        self.path = soundPath
    }

    function load_sound_sdl(soundPath: string, isMusic: boolean) {
        if (!(globalThis as any).__julgameAudioOpen) {
            return null
        }
        console.debug(`load_sound_sdl: Loading sound from ${soundPath}, isMusic: ${isMusic}`)
        if (haskey((globalThis as any).JulGame.AUDIO_CACHE, get_comma_separated_path(soundPath))) {
            let raw_data = (globalThis as any).JulGame.AUDIO_CACHE[get_comma_separated_path(soundPath)]
            let rw = (globalThis as any).JulGameSdl.glue_SDL_RWFromConstMem(pointer(raw_data), raw_data.length)
            if (rw != null) {
                console.debug("loading sound from cache")
                console.debug("comma separated path: ", get_comma_separated_path(soundPath))
                return isMusic ? (globalThis as any).JulGameSdl.glue_Mix_LoadMUS_RW(rw, 1) : (globalThis as any).JulGameSdl.glue_Mix_LoadWAV_RW(rw, 1)
            }
        }
        console.debug(`load_sound_sdl: Loading sound from disk, there are ${(globalThis as any).JulGame.AUDIO_CACHE.length} sounds in cache`)

        let fullPath = joinpath((globalThis as any).JulGame.BasePath, "assets", "sounds", soundPath)
        return isMusic ? (globalThis as any).JulGameSdl.glue_Mix_LoadMUS(fullPath) : (globalThis as any).JulGameSdl.glue_Mix_LoadWAV(fullPath)
    }

    function get_comma_separated_path(path: string) {
        // Normalize the path to use forward slashes
        let normalized_path = path.replace(/\\/g, '/')
        
        // Split the path into components
        let parts = normalized_path.split('/')
        
        let result = parts.join(",")
    
        return result  
    }

    function Component_unload_sound(self: InternalSoundSource) {
        console.debug(`unload_sound: Unloading sound from ${self.path}, isMusic: ${self.isMusic}`)
        if (self.isMusic) {
            (globalThis as any).JulGameSdl.glue_Mix_FreeMusic(self.sound)
        } else {
            (globalThis as any).JulGameSdl.glue_Mix_FreeChunk(self.sound)
        }
        self.sound = null
    }

    function Component_set_volume(self: InternalSoundSource, volume: number = 128, channel: number = -1) {
        console.debug(`set_volume: Setting volume for ${self.path}, isMusic: ${self.isMusic}, volume: ${volume}, channel: ${channel}`)
        // Convert volume to Int32 for SDL
        self.volume = mixVolume(volume)
        self.channel = clamp(channel, -1, 128)
        console.debug(`set_volume: Setting volume for ${self.path}, isMusic: ${self.isMusic}, volume: ${self.volume}, channel: ${self.channel}`)
        self.isMusic ? (globalThis as any).JulGameSdl.glue_Mix_VolumeMusic(mixVolume(self.volume)) : (globalThis as any).JulGameSdl.glue_Mix_Volume(self.channel, mixVolume(self.volume))
    }

    function Component_play(self: InternalSoundSource, loops: number = 0) {
        // Convert loops to Int32
        loops = loops
        
        console.debug(`play: Playing sound from ${self.path}, isMusic: ${self.isMusic}, channel: ${self.channel}, loops: ${loops}`)
        if (self.isMusic) {
            (globalThis as any).JulGameSdl.glue_Mix_VolumeMusic(mixVolume(self.volume));
            (globalThis as any).JulGameSdl.glue_Mix_PlayMusic(self.sound, -1)
        } else {
            (globalThis as any).JulGameSdl.glue_Mix_Volume(self.channel, mixVolume(self.volume));
            (globalThis as any).JulGameSdl.glue_Mix_PlayChannel(self.channel, self.sound, loops)
        }
    }

    function set_master_volume(volume: number) {
        // Convert volume to Int32 and clamp between 0 and 128
        console.debug(`set_master_volume: Setting master volume to ${volume}`)
        volume = mixVolume(volume);
        (globalThis as any).JulGameSdl.glue_Mix_MasterVolume(volume)
    }

    function Component_duplicate(self: InternalSoundSource, parent: any) {
        console.debug(`duplicate: Duplicating sound from ${self.path}, isMusic: ${self.isMusic}, channel: ${self.channel}, volume: ${self.volume}, playOnStart: ${self.playOnStart}`)
        let newSoundSource = new InternalSoundSource(parent, self.path, self.channel, self.volume, self.isMusic, self.playOnStart)
        newSoundSource.isPlaying = self.isPlaying
        return newSoundSource
    }
export { Component_duplicate, Component_load_sound, Component_play, Component_set_volume, Component_stop_music, Component_toggle_sound, Component_unload_sound, InternalSoundSource, SoundSource, get_comma_separated_path, load_sound_sdl, set_master_volume }
